privacyidea/lib/tokens/radiustoken.py

```diff
@@ -143,15 +144,193 @@ def update(self, param):
+        """
+        options = options or {}
+        message = options.get('radius_message') or "Enter your RADIUS tokencode:"
+        state = options.get('radius_state')
+        attributes = {'state': transactionid}
+        validity = int(get_from_config('DefaultChallengeValidityTime', 120))
+
+        log.debug("***[radius token: create challenge]")
+        log.debug("***loaded challenge {0}".format(state))
+
+        db_challenge = Challenge(self.token.serial,
+                                 transaction_id=transactionid,
+                                 data=state,
+                                 challenge=message,
+                                 validitytime=validity)
+        db_challenge.save()
+        self.challenge_janitor()
+        return True, message, db_challenge.transaction_id, attributes
+
+    def is_challenge_response(self, passw, user=None, options=None):
+        """
+        This method checks, if this is a request, that is the response to
+        a previously sent challenge.
+
+        Since this is RADIUS, we need to additionally check to see if the
+        current request is going to spawn another challenge
+
+        :param passw: password, which might be pin or pin+otp
+        :type passw: string
+        :param user: the requesting user
+        :type user: User object
+        :param options: dictionary of additional request parameters
+        :type options: dict
+        :return: true or false
+        :rtype: bool
+        """
+        options = options or {}
+        otpval = passw
+        challenge_response = False
+
+        log.debug("***[radius token: is challenge response]")
+
+        # clear the radius_result since this is the first function called in the chain
+        # this value will be utilized to ensure we do not check_radius more than once in the loop
+        options.update({'radius_result': None})
+
+        # fetch the transaction_id
+        transaction_id = options.get('transaction_id')
+        if transaction_id is None:
+            transaction_id = options.get('state')
+
+        if transaction_id:
+            # get the challenges for this transaction ID
+            challengeobject_list = get_challenges(serial=self.token.serial,
+                                                  transaction_id=transaction_id)
+
+            for challengeobject in challengeobject_list:
+                if challengeobject.is_valid():
+                    # challenge is still valid, pull the state and message and other data
+                    state = challengeobject.data
+
+                    # we need to check against radius before continuing
+                    # this causes a bit of an issue since we will know
+                    # if the challenge is successful before we try
+                    log.debug("***[radius token: challenge response state check] {0}".format(state))
+                    res = self.check_radius(otpval, options=options, radius_state=state)
```

Comment:
The method ``is_challenge_response`` only checks if the HTTP request is supposed to be a response to a previous challenge. So there is no need to verify with the remote RADIUS server.
The verification with the remote RADIUS server is supposed to be located in ``check_challenge_response``.

Reply:
So here's where I need your assistance on design.  The reason I'm doing this is due to the fact that the RADIUS server can challenge the user more than once. 

For example:
> Send OTP without PIN (initial setup/reset)
< Enter new PIN
> Send new PIN
< Re-enter new PIN
> Send new PIN again
< Enter next token code + PIN
> Send token+PIN
<< Auth-Success

I need the ability to trigger another challenge from within a challenge response.  I have basically written a hack/workaround that does a look-ahead within the is_challenge_response to see if the user is being challenged again or not.  If they are, I return False from is_challenge_response and True from is_challenge_request.

Is there a better way to do this?  I was hoping to not have to make changes to token.py if possible but it appears that may be a necessity.